### PR TITLE
[MNT] Distance test refactor

### PR DESCRIPTION
**Pull request by chrisholder**
#### What does this implement/fix? Explain your changes.
This PR refactors the distance tests to use the new testing utils for test time series generation. Specifically the '_create_test_distance_numpy' has been removed and where it was called replaced with make_series or make_example_3d_numpy which is consistent with the rest of the package.

#### Any other comments?
The test results have been updated because new time series were generated compared to the old.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [x] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed.

##### For developers with write access
- [x] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/distances/_distance.py [hunk 1]
```diff
@@ -138,7 +138,7 @@ def distance(
     elif metric == "manhattan":
         return manhattan_distance(x, y)
     elif metric == "minkowski":
-        return minkowski_distance(x, y)
+        return minkowski_distance(x, y, kwargs.get("p", 2.0), kwargs.get("w", None))
```
Comment (chrisholder) [inline aeon/distances/_distance.py:141]: I also spotted a bug and fixed the utils function call for minkowski_distance

File: aeon/distances/tests/test_pairwise.py [hunk 2]
```diff
@@ -99,29 +100,36 @@ def _validate_single_to_multiple_result(
...
-        _create_test_distance_numpy(5, 5),
+        make_series(5, 5, return_numpy=True, random_state=1),
         dist["name"],
         dist["distance"],
         dist["pairwise_distance"],
     )
 
     _validate_pairwise_result(
-        _create_test_distance_numpy(5, 1, 5),
+        make_example_3d_numpy(5, 1, 5, random_state=1)[0],
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:125]: I find this more explicit:
```suggestion
        make_example_3d_numpy(5, 1, 5, random_state=1, return_y=False),
```

File: aeon/distances/tests/test_pairwise.py [hunk 3]
```diff
@@ -142,59 +150,59 @@ def test_multiple_to_multiple_distances(dist):
...
         dist["distance"],
         dist["pairwise_distance"],
     )
 
     # Multivariate tests
     _validate_multiple_to_multiple_result(
-        _create_test_distance_numpy(5, 5),
-        _create_test_distance_numpy(5, 5, random_state=2),
+        make_series(5, 5, return_numpy=True, random_state=1),
+        make_series(5, 5, return_numpy=True, random_state=2),
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:163]: Yes, this creates two multivariate time series as input, but the `pairwise_distance`-functions do not support a single multivariate time series as 2-dim. ndarray. If a 2-dim. ndarray is passed, it is assumed to be a collection of univariate time series:

https://github.com/aeon-toolkit/aeon/blob/93726dc04caa2a83a61553100ac62c9cc8db9104/aeon/distances/_utils.py#L39-L42

File: aeon/distances/tests/test_pairwise.py [hunk 4]
```diff
@@ -218,58 +226,58 @@ def test_new_single_to_multiple_distances(dist):
...
 
     _validate_single_to_multiple_result(
-        _create_test_distance_numpy(5),
-        _create_test_distance_numpy(3, 1, 5, random_state=2),
+        make_series(5, 1, return_numpy=True, random_state=1),
+        make_example_3d_numpy(3, 1, 5, random_state=2)[0],
         dist["name"],
         dist["distance"],
         dist["pairwise_distance"],
     )
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:234]: We should also test for:

```python
X = make_example_3d_numpy(3, 1, 5, random_state=2)[0]
y = make_series(5, 1, return_numpy=True, random_state=1)
```

File: aeon/distances/tests/test_pairwise.py [hunk 5]
```diff
@@ -218,58 +226,58 @@ def test_new_single_to_multiple_distances(dist):
...
 
     _validate_single_to_multiple_result(
-        _create_test_distance_numpy(3, 1, 5)[0],
-        _create_test_distance_numpy(5, 1, 5, random_state=2),
+        make_series(1, 5, return_numpy=True, random_state=1),
+        make_example_3d_numpy(5, 1, 5, random_state=2)[0],
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:238]: This creates:

```python
X.shape = (5, 1)    # --reshaped--> (5, 1, 1)
y.shape = (5, 1, 5) # --reshaped--> (5, 1, 5)
```

It is not obvious to me, why we need this test, and why you use `make_series` to create `X`.

File: aeon/distances/tests/test_pairwise.py [hunk 6]
```diff
@@ -218,58 +226,58 @@ def test_new_single_to_multiple_distances(dist):
...
         dist["distance"],
         dist["pairwise_distance"],
     )
 
     # Multivariate tests
     _validate_single_to_multiple_result(
-        _create_test_distance_numpy(5, 5),
-        _create_test_distance_numpy(5, 5, 5, random_state=2),
+        make_series(5, 5, return_numpy=True, random_state=1),
+        make_example_3d_numpy(5, 5, 5, random_state=2)[0],
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:247]: Again, `X` is not a multivariate time series, but interpreted as `(5, 1, 5)`. This is compared to `(5, 5, 5)` testing for multivariate time series with different number of channels.

File: aeon/distances/tests/test_pairwise.py [hunk 7]
```diff
@@ -218,58 +226,58 @@ def test_new_single_to_multiple_distances(dist):
...
         dist["name"],
         dist["distance"],
         dist["pairwise_distance"],
     )
 
     _validate_single_to_multiple_result(
-        _create_test_distance_numpy(5),
-        _create_test_distance_numpy(3, 1, 3, random_state=2),
+        make_series(3, return_numpy=True, random_state=1),
+        make_example_3d_numpy(3, 1, 3, random_state=2)[0],
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:264]: This is an equal-length test `(1, 1, 3)` vs. `(3, 1, 3)`.

File: aeon/distances/tests/test_pairwise.py [hunk 8]
```diff
@@ -218,58 +226,58 @@ def test_new_single_to_multiple_distances(dist):
...
         dist["name"],
         dist["distance"],
         dist["pairwise_distance"],
     )
 
     _validate_single_to_multiple_result(
-        _create_test_distance_numpy(5, 3),
-        _create_test_distance_numpy(5, 5, 5, random_state=2),
+        make_series(5, 3, return_numpy=True, random_state=1),
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:271]: To make it an unequal-length test:
```suggestion
        make_series(3, 5, return_numpy=True, random_state=1),
```

File: aeon/distances/tests/test_pairwise.py [hunk 9]
```diff
@@ -4,7 +4,8 @@
 
 from aeon.distances import pairwise_distance as compute_pairwise_distance
 from aeon.distances._distance import DISTANCES
-from aeon.distances.tests.test_utils import _create_test_distance_numpy, _make_3d_series
+from aeon.distances.tests.test_utils import _make_3d_series
```
Comment (CodeLionX) [inline aeon/distances/tests/test_pairwise.py:7]: Do we still need `_make_3d_series` or can we replace it as well?

File: aeon/distances/tests/test_pairwise.py (hunk 2 above)
Comment (TonyBagnall) [reply aeon/distances/tests/test_pairwise.py:125]: I agree with @CodeLionX :)

Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py:125]: So this maybe a crazy reason for doing this: The type hint for the function indicated it returns a tuple always and it was setting off my linter. I will just the type hint of the function as well and add return_y=False.

File: aeon/distances/tests/test_pairwise.py (hunk 3 above)
Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py:163]: The intention of passing all of these in different formats is hitting the logic inside pairwise that does the reshaping:

```
    if y is None:
        # To self
        if X.ndim == 3:
            return _msm_pairwise_distance(X, window, independent, c, itakura_max_slope)
        if X.ndim == 2:
            _X = X.reshape((X.shape[0], 1, X.shape[1]))
            return _msm_pairwise_distance(_X, window, independent, c, itakura_max_slope)
        raise ValueError("x and y must be 2D or 3D arrays")
    elif y.ndim == X.ndim:
        # Multiple to multiple
        if y.ndim == 3 and X.ndim == 3:
            return _msm_from_multiple_to_multiple_distance(
                X, y, window, independent, c, itakura_max_slope
            )
        if y.ndim == 2 and X.ndim == 2:
            _x = X.reshape((X.shape[0], 1, X.shape[1]))
            _y = y.reshape((y.shape[0], 1, y.shape[1]))
            return _msm_from_multiple_to_multiple_distance(
                _x, _y, window, independent, c, itakura_max_slope
            )
        if y.ndim == 1 and X.ndim == 1:
            _x = X.reshape((1, 1, X.shape[0]))
            _y = y.reshape((1, 1, y.shape[0]))
            return _msm_from_multiple_to_multiple_distance(
                _x, _y, window, independent, c, itakura_max_slope
            )
        raise ValueError("x and y must be 1D, 2D, or 3D arrays")
```

Here im just using make_series as a way to create a collection univariate (even though I know the function is technically creating a multivariate series). I'll add comments for clarity.

Comment (CodeLionX) [reply aeon/distances/tests/test_pairwise.py:163]: I get your intention, but I don't know if `make_series` is a good way of achieving this. At least a brief explanation would help with understanding the tests, if we stick to `make_series` for 2d-ndarray input, agreed 👍🏼

File: aeon/distances/tests/test_utils.py [hunk 10]
```diff
@@ -48,3 +33,42 @@ def test_reshape_pairwise_to_multiple():
...
+
+    Pairwise assumes it has been passed two sets of series, if passed a single
+    series this function reshapes.
+
+    If given a 1d array the time series is reshaped to (m, 1, 1). This is so when
+    looped over x[i] = (1, m).
+
+    If given a 2d array then the time series is reshaped to (d, 1, m). The dimensions
+    are put to the start so the ts can be looped through correctly. When looped over
+    the time series x[i] = (d, m).
```
Comment (CodeLionX) [inline aeon/distances/tests/test_utils.py:49]: Isn't it rather like this?
```suggestion
    If given a 1d array the time series is reshaped to (1, 1, m). This is so when
    looped over x[i] = (1, m).

    If given a 2d array then the time series is reshaped to (d, 1, m). The dimensions
    are put to the start so the ts can be looped through correctly. When looped over
    the time series x[i] = (1, m).
```

File: aeon/distances/tests/test_distances.py [hunk 11]
```diff
@@ -12,10 +12,15 @@
...
 )
 from aeon.distances.tests.test_expected_results import _expected_distance_results
-from aeon.distances.tests.test_utils import _create_test_distance_numpy
+from aeon.distances.tests.test_utils import SINGLE_POINT_NOT_SUPPORTED_DISTANCES
+from aeon.testing.utils.data_gen import make_series
 
 
-def _validate_distance_result(x, y, name, distance, expected_result=10):
+def _validate_distance_result(
+    x, y, name, distance, expected_result=10, recursive_call=False
```
Comment (CodeLionX) [inline aeon/distances/tests/test_distances.py:20]: I would rename `recursive_call` to `check_xy_permuted` or similar to express the intention behind the _recursion_.

File: aeon/distances/tests/test_distances.py [hunk 12]
```diff
@@ -29,53 +34,91 @@ def _validate_distance_result(x, y, name, distance, expected_result=10):
...
+    # ================== Test unequal length ==================
+    # Test univariate unequal length of shape (n_timepoints,)
     _validate_distance_result(
-        _create_test_distance_numpy(5),
-        _create_test_distance_numpy(10, random_state=2),
+        make_series(5, return_numpy=True, random_state=1),
+        make_series(10, return_numpy=True, random_state=2),
         dist["name"],
         dist["distance"],
         _expected_distance_results[dist["name"]][3],
```
Comment (CodeLionX) [inline aeon/distances/tests/test_distances.py:81]: Did you forget to enable `recursive_call` (`check_xy_permuted`) for the unequal-length tests?

File: aeon/distances/tests/test_alignment_path.py [hunk 13]
```diff
@@ -4,16 +4,20 @@
...
 
 
-def _validate_cost_matrix_result(
+def _validate_alignment_path_result(
     x: np.ndarray,
     y: np.ndarray,
     name,
     distance,
     alignment_path,
+    recursive_call=False,
```
Comment (CodeLionX) [inline aeon/distances/tests/test_alignment_path.py:17]: Again, not used. Did you forget to enable it?

**Review by CodeLionX — CHANGES_REQUESTED** [review]
Nice work and way more clearer tests than before 👍🏼 
I guess, you forgot to actually use/enable the recursive option for all unequal-length tests. I left comments hinting at this just for some files.

File: aeon/distances/tests/test_pairwise.py [hunk 14]
```diff
@@ -99,29 +100,36 @@ def _validate_single_to_multiple_result(
...
+        dist = distance(curr_x, curr_y)
         assert_almost_equal(dist, curr)
 
 
 @pytest.mark.parametrize("dist", DISTANCES)
 def test_pairwise_distance(dist):
     """Test pairwise distance function."""
     _validate_pairwise_result(
-        _create_test_distance_numpy(5, 5),
+        make_series(5, 5, return_numpy=True, random_state=1),
```
[the remark this answers is not in the document]
Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py]: Talked on slack about this and the new changes resolve this

File: aeon/distances/tests/test_pairwise.py (hunk 9 above)
Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py]: There are tests that need this to convert into a known format that works. For now will keep but maybe will removed in the near future. I have moved it into test utils as it is the only place it is used.

File: aeon/distances/tests/test_pairwise.py (hunk 6 above)
Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py]: Fixed in update

File: aeon/distances/tests/test_pairwise.py (hunk 5 above)
Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py]: Fixed in update

File: aeon/distances/tests/test_pairwise.py (hunk 4 above)
Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py]: Added in update

File: aeon/distances/tests/test_pairwise.py (hunk 3 above)
Comment (chrisholder) [reply aeon/distances/tests/test_pairwise.py]: Fixed in new update and talked further on slack

File: aeon/distances/tests/test_cost_matrix.py [hunk 15]
```diff
@@ -13,7 +14,10 @@ def _validate_cost_matrix_result(
     name,
     distance,
     cost_matrix,
+    recursive_call=False,
```
[the remark this answers is not in the document]
Comment (chrisholder) [reply aeon/distances/tests/test_cost_matrix.py:17]: It's snuck in at the end of line 51

File: aeon/distances/tests/test_utils.py (hunk 10 above)
Comment (chrisholder) [reply aeon/distances/tests/test_utils.py:49]: Excellent spot yes!

File: aeon/distances/tests/test_alignment_path.py [hunk 16]
```diff
@@ -4,16 +4,20 @@
...
 
 
-def _validate_cost_matrix_result(
+def _validate_alignment_path_result(
     x: np.ndarray,
     y: np.ndarray,
     name,
     distance,
     alignment_path,
+    check_xy_permuted=False,
```
Comment (CodeLionX) [inline aeon/distances/tests/test_alignment_path.py:17]:
```suggestion
    check_xy_permuted=True,
```

File: aeon/distances/tests/test_alignment_path.py [hunk 17]
```diff
@@ -24,52 +28,98 @@ def _validate_cost_matrix_result(
...
+    # If unequal length swap where x and y are to ensure it works both ways around
+    if original_x.shape[-1] != original_y.shape[-1] and not check_xy_permuted:
+        _validate_alignment_path_result(
+            original_y,
+            original_x,
+            name,
+            distance,
+            alignment_path,
+            check_xy_permuted=True,
+        )
```
Comment (CodeLionX) [inline aeon/distances/tests/test_alignment_path.py:40]:
```suggestion
    if original_x.shape[-1] != original_y.shape[-1] and check_xy_permuted:
        _validate_alignment_path_result(
            original_y,
            original_x,
            name,
            distance,
            alignment_path,
            check_xy_permuted=False,
        )
```

**Review by CodeLionX — COMMENTED** [review]
This would be clearer to me.

File: aeon/distances/tests/test_alignment_path.py (hunk 16 above)
Comment (chrisholder) [reply aeon/distances/tests/test_alignment_path.py:17]: I agree

**Review by CodeLionX — APPROVED** [review]
This is a great refactoring of the distance tests; no more comments from my side! 🚀 I'm really looking forward to basing my additions on this in #1287

**Review by TonyBagnall — APPROVED** [review]

**Pull request merged**